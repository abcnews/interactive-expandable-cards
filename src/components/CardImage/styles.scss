.image {
  margin: 0.5rem -0.25rem 0;

  @media (min-width: 40rem) {
    margin: 0.75rem 0 0;
  }

  &:first-child {
    margin-top: 0;

    .siblingsHaveLabels > & {
      margin-top: 1.5rem;

      @media (min-width: 40rem) {
        margin-top: 1.75rem;
      }
    }
  }

  & > img {
    max-width: 100%;
    height: auto;
    vertical-align: top;
  }
}
