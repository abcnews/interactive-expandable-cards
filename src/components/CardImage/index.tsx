import { h, FunctionalComponent } from 'preact';
import { useEffect, useMemo, useState } from 'preact/hooks';
import styles from './styles.scss';
import GammaFilter from '../GammaFilter';
import { RGB, rgbGamma } from '../../lib/utils';
import { ExpandableCardsImage } from '../ExpandableCards';

let idCounter = 0;

type CardImageProps = {
  image: ExpandableCardsImage;
  tint: false | RGB;
};

const CardImage: FunctionalComponent<CardImageProps> = ({ image, tint }) => {
  const [srcSet, setSrcSet] = useState<string[]>([]);
  const id = useMemo<string>(() => '' + idCounter++, []);

  const filterId = `ExpandableCards__Filter_${id}`;
  const filterCSS = tint ? `grayscale(100%) url(#${filterId})` : undefined;

  useEffect(() => {
    setSrcSet(image.renditions.map(i => `${i.url} ${i.width}w`));
  }, [image]);

  return (
    <figure className={styles.image} role="presentation">
      <img
        alt={image.alt}
        src={image.url}
        srcset={srcSet.join(', ')}
        sizes="(min-width: 1200px) 180px, (max-width: 1200px; min-width: 992px) 145px, 210px"
        style={{ filter: filterCSS, '-webkit-filter': filterCSS }}
      />

      {tint && <GammaFilter id={filterId} exponents={rgbGamma(tint)} />}
    </figure>
  );
};

export default CardImage;
