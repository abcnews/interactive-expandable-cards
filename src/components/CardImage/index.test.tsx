import { h } from 'preact';
import render from 'preact-render-to-string';

import CardImage from '.';

describe('CardImage', () => {
  test('It renders', () => {
    const actual = render(<CardImage image={{ alt: 'Alt', url: 'url', renditions: [] }} tint={false} />);
    expect(actual).toMatchSnapshot();
  });
});
