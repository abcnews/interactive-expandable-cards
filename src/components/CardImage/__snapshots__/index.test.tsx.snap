// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardImage It renders 1`] = `"<figure role=\\"presentation\\"><img alt=\\"Alt\\" src=\\"url\\" srcset sizes=\\"(min-width: 1200px) 180px, (max-width: 1200px; min-width: 992px) 145px, 210px\\"/></figure>"`;
